### ISSUE TITLE
validate_url and download_extension request every URL twice, and don't re-use connections, making builds really slow

### ISSUE THREAD

**Issue by miceg** [kartoza/docker-geoserver#642]
### What is the bug or the crash?

`validate_url` and `download_extension` request every URL twice:

https://github.com/kartoza/docker-geoserver/blob/2d02b4df3e26df1cbb64b3cb4757e2f7aed597a2/scripts/functions.sh#L15-L16

https://github.com/kartoza/docker-geoserver/blob/939ee005d27fe123ce75c832e9706fe2825dc642/scripts/functions.sh#L129-L130

1. They first make a HTTP head request, and check the server response code, but in different ways:
   * `validate_url` fetches the URL with `wget -S --spider`
   * `download_extension` fetches the URL with `curl --head`
2. The URL is fetched with `wget` normally, and saved to a file.

As a shell script, there's no connection re-use between `wget`/`curl` calls, so this goes through a TCP (and TLS) handshake twice. That's 1 round-trip for HTTP or 3 round-trips for HTTPS... before you've even sent the HTTP request (another round-trip).

Many requests go to the same host (eg: `build.geoserver.org`, `sourceforge.net`), so could benefit by passing a URL list to `wget` rather than starting up new processes.

### Steps to reproduce the issue

Build the image with any significant amount of latency to `build.geoserver.org` or other sites – it takes many minutes.

### Versions

Current develop branch: https://github.com/kartoza/docker-geoserver/commit/2d02b4df3e26df1cbb64b3cb4757e2f7aed597a2

### Additional context

I'm sitting at about 200 ms latency to `build.geoserver.org`, so each HTTPS request takes at least 800ms (3 round-trips setup + 1 round-trip for the actual HTTP request) before the server has even sent a single byte of content. Doing that *twice* blows it out to 1.6 seconds/URL, and that's just network latency.

Looking at *just* the community plugins, with 68 community plugins, sending just *one* request per plugin would reduce build time by at least 54 seconds (at 200 ms latency).

If the download code was refactored so that all the plugin download requests happened in the same `wget` process, it could benefit from connection re-use, and get off 3 round-trips for every URL after the first (204 round-trips total = 41 seconds).

All up, fixing *just* the community plugins would reduce the build time at least 1.5 minutes (at 200 ms latency). I suspect the actual savings (for *all* downloads) would be much higher, and higher your latency, the bigger the impact.

Docker cache mounts (`RUN --mount=type=cache`) would help here too, but only for re-builds. That'd avoid fetching URLs entirely the client already has.

**Comment by miceg** [comment]
I've implemented this in a private repo, I added a separate layer which generates a `curl` configuration file to download everything, and runs that. It runs in parallel with the rest of the `geoserver-prod` image, so there's a small speed-up there too.

With my changes, clean build times went from 8min 19sec to 4min 18sec. Rebuilds with no plugin changes are even faster: around 2 minutes!

Docker cache mounts would make changing the plugin set faster again, but would require more architectural changes.

It would be even faster, but the "required" and "stable" GeoServer plugin downloads are on SourceForge, which takes about 1.5 - 3.0 sec per URL to bounce through their terrible redirect chain, which is across three different hostnames... and even though they use a CDN. 🙃 

Upstreaming this change is currently blocked on https://github.com/kartoza/docker-geoserver/pull/651, because that PR changes the set of required plugins, and my implementation also incorporates this change.
